urld2fb1d2b05_thnaeff_Util_tgz-pJ8-ch_thn_util_tests_TreeTestJ8/src/ch/thn/util/collections/MultiValueHashMap.java

```diff
@@ -0,0 +1,325 @@
+ * thus the order in which the values are added is also the order in which they 
+ * are stored.
+ * 
+ * @author Thomas Naeff (github.com/thnaeff)
+ *
+ */
+public class MultiValueHashMap<K, V> extends AbstractMap<K, V> {
+	
+	protected Map<K, LinkedList<V>> map = null;
+	
+	protected int size = 0;
+	
+	/**
+	 * 
+	 * @see HashMap#HashMap()
+	 */
+	public MultiValueHashMap() {
+		map = new HashMap<K, LinkedList<V>>();
+	}
+	
+	
+	/**
+	 * 
+	 * 
+	 * @param initialCapacity
+	 * @see HashMap#HashMap(int)
+	 */
+	public MultiValueHashMap(int initialCapacity) {
+		map = new HashMap<K, LinkedList<V>>(initialCapacity);
+	}
+	
+	/**
+	 * 
+	 * 
+	 * @param m
+	 * @see HashMap#HashMap(Map)
+	 */
+	public MultiValueHashMap(Map<? extends K, ? extends V> m) {
+		map = new HashMap<K, LinkedList<V>>();
+		putAll(m);
+	}
+	
+	/**
+	 * 
+	 * 
+	 * @param initialCapacity
+	 * @param loadFactor
+	 * @see HashMap#HashMap(int, float)
+	 */
+	public MultiValueHashMap(int initialCapacity, float loadFactor) {
+		map = new HashMap<K, LinkedList<V>>(initialCapacity, loadFactor);
+	}
+	
+	/**
+	 * Associates the specified value with the specified key in this map (optional operation). 
+	 * If the map previously contained a mapping for the key, the new value 
+	 * is added to the value list for that key.
+	 * 
+	 * @param key
+	 * @param value
+	 * @return the previous value associated with key, or null if there was no 
+	 * mapping for key (A null return can also indicate that the map previously 
+	 * associated null with key, if the implementation supports null values). This 
+	 * means for this {@link MultiValueHashMap}, the put-methods 
+	 * return the last value which has been added to the list for the given key.
+	 */
+	@Override
+	public V put(K key, V value) {
+		if (!map.containsKey(key)) {
+			map.put(key, new LinkedList<V>());
+		}
+		
+		V previous = null;
+		if (map.get(key).size() > 0) {
+			previous = map.get(key).getLast();
+		}
+		
+		map.get(key).add(value);
+		size++;
+		
+		/* JavaDoc for Map.put(K, V):
+		 * Returns: the previous value associated with key, or null if there was 
+		 * no mapping for key. (A null return can also indicate that the map 
+		 * previously associated null with key, if the implementation supports 
+		 * null values.)
+		 */
+		return previous;
+	}
+	
+	@Override
+	public void putAll(Map<? extends K, ? extends V> m) {
+		try {
+			if (m instanceof MultiValueHashMap) {
+				@SuppressWarnings("unchecked")
+				MultiValueHashMap<K, V> mvhm = (MultiValueHashMap<K, V>)m;
+				for (K key : mvhm.keySet()) {
+					//Add the content list by list
+					if (!map.containsKey(key)) {
+						map.put(key, new LinkedList<V>(mvhm.getAll(key)));
+					} else {
+						map.get(key).addAll(mvhm.getAll(key));
+					}
+					size += map.get(key).size();
+				}
+			}
+		} catch (ClassCastException e) {
+			//If there is a class cast exception, just add the content individually
+		}
+		
+		for (K key : m.keySet()) {
+			put(key, m.get(key));
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**: Iterating over [map.keySet()](https://docs.oracle.com/javase/8/docs/api/java/util/Map.html#keySet--) and calling [map.get(Key)](https://docs.oracle.com/javase/8/docs/api/java/util/Map.html#get-java.lang.Object-) to get the values within the loop is inefficient. 
**Fix**: Iterate over [map.entrySet()](https://docs.oracle.com/javase/8/docs/api/java/util/Map.html#entrySet--) and access the values using [entry.getValue()](https://docs.oracle.com/javase/8/docs/api/java/util/Map.Entry.html#getValue--) to avoid potential performance penalty due to lookup.